spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/admin/SpannerDatabaseAdminTemplate.java

```diff
@@ -54,32 +55,31 @@
 
 	private final DatabaseAdminClient databaseAdminClient;
 
-	private final DatabaseId databaseId;
+	private final Supplier<DatabaseId> databaseIdProvider;
 
-	private final DatabaseClient databaseClient;
+	private final Supplier<DatabaseClient> databaseClientProvider;
 
 	/**
-	 * Constructor that takes in the database admin client used to perform operations and
-	 * the {@link DatabaseId} object holding the project, instance, and database IDs used
-	 * for all operations. While operations can be optionally performed for a database
-	 * that does not yet exist, the project and instance IDs must already exist for
-	 * Spanner.
+	 * Constructor that takes in the database admin client used to perform operations and the
+	 * {@link DatabaseId} object holding the project, instance, and database IDs used for all
+	 * operations. While operations can be optionally performed for a database that does not
+	 * yet exist, the project and instance IDs must already exist for Spanner.
 	 * @param databaseAdminClient the client used to create databases and execute DDL
-	 * statements.
-	 * @param databaseClient the client used to access schema information tables.
-	 * @param databaseId the combination of Cloud Spanner Instance Id and Database Id. While
-	 * databases can be created automatically by this template, instances determine
-	 * billing and are not created automatically.
+	 *     statements.
+	 * @param databaseClientProvider the client used to access schema information tables.
+	 * @param databaseIdProvider the combination of Cloud Spanner Instance Id and Database Id.
+	 *     While databases can be created automatically by this template, instances determine
+	 *     billing and are not created automatically.
 	 */
 	public SpannerDatabaseAdminTemplate(DatabaseAdminClient databaseAdminClient,
-			DatabaseClient databaseClient,
-			DatabaseId databaseId) {
+										Supplier<DatabaseClient> databaseClientProvider,
```

Comment:
Why would you need this, if you can compute it from the `databaseIdProvider`?

Reply:
If the user had chosen to provide his own DatabaseClient supplier then that wouldn't be right (for example if he sets things that aren't in DatabaseId). Also if we just use the DAtabaseIdProvider we would need to create new instances of DatabaseClient in here and also have a Spanner.class on hand.